### PR TITLE
[sc-104229] implement batch on event frames

### PR DESCRIPTION
**Pull request by alexbourret**

### CODE REVIEW

**Review by liamlynch-data — COMMENTED** [review]
I have added various comments/suggestions to consider but I don't think there are any big problems, in general it seems fine. 

I did testing on the a PR/branch further down the chain (bug/sc-114420-nan-sync-transposed-dataset) as some of the issues I noticed were fixed there. But there was still a problem if you choose batch mode and data type = None - but I think you are fixing that (e.g. by removing that None option)? Other than that it seems to work fine.

File: python-lib/osisoft_client.py [hunk 1]
```diff
@@ -63,6 +63,60 @@ def get_row_from_webid(self, webid, data_type, start_date=None, end_date=None,
...
+            if (len(batch) >= batch_size) or (number_processed_webids == number_of_webids_to_process):
+                json_responses = self.process_batch(batch)
+                for json_response in json_responses:
+                    if OSIsoftConstants.DKU_ERROR_KEY in json_response:
+                        json_response['object_id'] = "{}".format(webid.get("WebId"))
+                        yield json_response
+                    items = json_response.get(OSIsoftConstants.API_ITEM_KEY, [])
+                    for item in items:
+                        yield item
+                batch = []
```
[the remark this answers is not in the document]
Comment (alexbourret) [reply python-lib/osisoft_client.py:87]: Fixed in [a83525e](https://github.com/dataiku/dss-plugin-osisoft/pull/20/commits/a83525e530607e1fc0beef7a0a29bf4d24130a73)

File: python-connectors/pi-system_event-frames-search/connector.py [hunk 2]
```diff
@@ -61,39 +65,60 @@ def generate_rows(self, dataset_schema=None, dataset_partitioning=None,
...
-                            yield event_frame_copy
+            if self.must_retrieve_metrics:
+                if use_batch_mode:
+                    batch_rows = self.client.get_rows_from_webids(
+                            event_frames, self.data_type,
+                            can_raise=False,
+                            batch_size=self.batch_size
+                        )
+                    for batch_row in batch_rows:
+                        items = batch_row.pop("Items", [])
```
[the remark this answers is not in the document]
Comment (alexbourret) [reply python-connectors/pi-system_event-frames-search/connector.py:76]: Fixed in [a83525e](https://github.com/dataiku/dss-plugin-osisoft/pull/20/commits/a83525e530607e1fc0beef7a0a29bf4d24130a73)

File: python-lib/osisoft_client.py [hunk 3]
```diff
@@ -63,6 +63,60 @@ def get_row_from_webid(self, webid, data_type, start_date=None, end_date=None,
...
+
+    def process_batch(self, batch):
```
[the remark this answers is not in the document]
Comment (alexbourret) [reply python-lib/osisoft_client.py:89]: Fixed in [a83525e](https://github.com/dataiku/dss-plugin-osisoft/pull/20/commits/a83525e530607e1fc0beef7a0a29bf4d24130a73)

File: python-lib/osisoft_client.py [hunk 4]
```diff
@@ -63,6 +63,60 @@ def get_row_from_webid(self, webid, data_type, start_date=None, end_date=None,
             for item in items:
                 yield item
 
+    def get_rows_from_webids(self, webids, data_type, start_date=None, end_date=None,
+                           interval=None, sync_time=None, boundary_type=None, selected_fields=None,
+                           can_raise=True, endpoint_type="event_frames", batch_size=500):
+        batch = []
```
[the remark this answers is not in the document]
Comment (alexbourret) [reply python-lib/osisoft_client.py:69]: Fixed in [a83525e](https://github.com/dataiku/dss-plugin-osisoft/pull/20/commits/a83525e530607e1fc0beef7a0a29bf4d24130a73)

[the remark this answers is not in the document]
Comment (liamlynch-data) [reply python-lib/osisoft_client.py:69]: great!

File: python-connectors/pi-system_event-frames-search/connector.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (liamlynch-data) [reply python-connectors/pi-system_event-frames-search/connector.py:76]: 👍

File: python-lib/osisoft_client.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (liamlynch-data) [reply python-lib/osisoft_client.py:87]: Actually I misunderstood this bit of code before, and the line web_ids = [] has to be at the end under the if statement - nevermind!

File: python-connectors/pi-system_event-frames-search/connector.py [hunk 5]
```diff
@@ -61,39 +65,60 @@ def generate_rows(self, dataset_schema=None, dataset_partitioning=None,
...
+                        for item in items:
+                            batch_row.update(item)
+                            yield batch_row
                             limit.add_record()
+                            if limit.is_reached():
+                                break
```
[the remark this answers is not in the document]
Comment (liamlynch-data) [reply python-connectors/pi-system_event-frames-search/connector.py:82]: was only a suggestion - deemed to much effort/risk at this stage

File: python-connectors/pi-system_event-frames-search/connector.py [hunk 6]
```diff
@@ -61,39 +65,60 @@ def generate_rows(self, dataset_schema=None, dataset_partitioning=None,
...
+                                    row.update(item)
+                                    row.pop("Links", None)
+                                    yield row
+                                    limit.add_record()
+                            else:
+                                event_frame_copy.pop("Links", None)
+                                yield event_frame_copy
+                                limit.add_record()
+                        if limit.is_reached():
+                            break
```
[the remark this answers is not in the document]
Comment (liamlynch-data) [reply python-connectors/pi-system_event-frames-search/connector.py:108]: was only a suggestion - deemed to much effort/risk at this stage

**Review by liamlynch-data — APPROVED** [review]
The issues I was worried about have either been resolved in upstream PRs to the same release version or we considered them not worth changing now - so approving

**Pull request merged**